bigchaindb/backend/mongodb/connection.py

```diff
@@ -71,22 +75,29 @@ def _connect(self):
             # we should only return a connection if the replica set is
             # initialized. initialize_replica_set will check if the
             # replica set is initialized else it will initialize it.
-            initialize_replica_set(self.host, self.port, self.connection_timeout)
+            initialize_replica_set(self.host, self.port, self.connection_timeout,
+                                   self.dbname, self.ssl, self.login, self.password)
 
             # FYI: this might raise a `ServerSelectionTimeoutError`,
             # that is a subclass of `ConnectionFailure`.
-            return pymongo.MongoClient(self.host,
-                                       self.port,
-                                       replicaset=self.replicaset,
-                                       serverselectiontimeoutms=self.connection_timeout)
+            client = pymongo.MongoClient(self.host,
+                                         self.port,
+                                         replicaset=self.replicaset,
+                                         serverselectiontimeoutms=self.connection_timeout,
+                                         ssl=self.ssl)
+
+            if self.login is not None and self.password is not None:
+                client[self.dbname].authenticate(self.login, self.password)
+
+            return client
 
         # `initialize_replica_set` might raise `ConnectionFailure` or `OperationFailure`.
         except (pymongo.errors.ConnectionFailure,
                 pymongo.errors.OperationFailure) as exc:
             raise ConnectionError() from exc
```

Comment:
Unfortunately `pymongo` doesn't provide a specific exception when the authentication fails, but a generic `OperationFailure`. This might trigger a retry, but it's not the correct behavior: if credentials are wrong we want BigchainDB to fail and give an explicit error message to the user.

You **don't have to care about** it in this PR, some of us can do it after we merge your PR :smile:

See #1336